### ISSUE TITLE
A case not handled by `no-for-loop`

### ISSUE THREAD

**Issue by chiawendt** [sindresorhus/eslint-plugin-unicorn#295]
[<img alt="Issuehunt badges" src="https://img.shields.io/badge/IssueHunt-%2440%20Funded-%2300A156.svg" />](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/295)

```js
/* eslint unicorn/no-for-loop: 2 */

const arr = [];

for (let i = 0, j = arr.length; i < j; i += 1) {
  const element = arr[i];
  console.log(element);
}
```
`no-for-loop` should report on this but it does not.

This slightly unusual pattern is used all over the place in `eslint-plugin-react`.

This issue is a subtask of #250.

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Sponsors (Total: $40.00)

- [<img src='https://avatars3.githubusercontent.com/u/44827199?v=4' alt='issuehunt' width=24 height=24> issuehunt](https://issuehunt.io/u/issuehunt) ($40.00)

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
</details>

**Comment by sindresorhus** [comment]
@golopot This was not handled by https://github.com/sindresorhus/eslint-plugin-unicorn/pull/297, right?

**Comment by golopot** [comment]
It is not.

**Comment by IssueHuntBot** [comment]
@issuehunt has funded $40.00 to this issue.

---
- Submit pull request via [IssueHunt](https://issuehunt.io/repos/55832243/issues/295) to receive this reward.
- Want to contribute? Chip in to this issue via [IssueHunt](https://issuehunt.io/repos/55832243/issues/295).
- Checkout the [IssueHunt Issue Explorer](https://issuehunt.io/issues) to see more funded issues.
- Need help from developers? [Add your repository](https://issuehunt.io/r/new) on IssueHunt to raise funds.

**Comment by MrHen** [comment]
I took a quick look at this and think this should be a separate rule that auto-fixes this shape into a normal for loop. `no-for-loop` can then convert the normal for loop since it already handles that.

Original:

```
for (let i = 0, j = arr.length; i < j; i += 1) {
  const element = arr[i];
  console.log(element);
}
```

Fixed:

```
for (let i = 0; i < arr.length; i += 1) {
  const element = arr[i];
  console.log(element);
}
```

----

Original:

```
for (let i = 0, j = arr.length; i < j; i += 1) {
  const element = arr[i];
  console.log(element);
  console.log(j);
}
```

Fixed:

```
for (let i = 0; i < arr.length; i += 1) {
  const element = arr[i];
  console.log(element);
  console.log(arr.length);
}
```

**Comment by sindresorhus** [comment]
@MrHen Good point. Are we sure we can depend on the fixer output of other rules (the order of fixes)? I don't fully remember how fixers work.

How about a rule named `simplify-for-loop`?

**Comment by MrHen** [comment]
@sindresorhus `auto-fix` will repeatedly attempt to validate after fixes. I'm not sure what happens if two things want to fix the same code but I don't think it would matter here because the fixers are not going to immediately overlap.

The main reason I'd want this check in a different rule is because it would make the rule logic and test cases easier to understand and maintain. The end result should be:

```
for (let i = 0, j = arr.length; i < j; i += 1) {
  const element = arr[i];
  console.log(element);
}
```

Will be auto-fixed to:

```
for (let i = 0; i < arr.length; i += 1) {
  const element = arr[i];
  console.log(element);
}
```

Which is then auto fixed to:

```
for (const [i, element] of arr) {
  console.log(element);
}
```

The end result will be exactly what you want.

**Comment by sindresorhus** [comment]
Alright, let's do it as a separate rule then.

**Comment by jmoore914** [comment]
I can take this one!

**Comment by fisker** [comment]
@jmoore914 I thought this is fixed by #476 , but apparently `ArrayPattern` is missed, I just send a follow up PR #489 fix this.

@sindresorhus @MrHen I missed this issue, when I review #476, we already do destructuring in the `for...of`.

**Comment by snowteamer** [comment]
> Alright, let's do it as a separate rule then.

What about restricting the scope of this proposed new separate rule to only address array length caching ?
- it would still address the specific OP issue
- simpler than checking for every possible way of declaring an extra variable
- array length caching was once a possible performance optimization, but nowadays modern browsers automatically cache it in such for loops, making manual caching an obsolete practice unless targeting old browsers
- all proposed pass/fail exemples in #492 already concern a "length" variable anyway
- maybe also the possibility to name the new rule `no-array-length-caching` instead of `no-for-loop-extra-variable`

**Comment by sindresorhus** [comment]
If anyone wants to work on this, please continue where https://github.com/sindresorhus/eslint-plugin-unicorn/pull/492 left off.

**Comment by d-ship-it26** [comment]
I have a fix ready for this issue here:

https://github.com/d-ship-it26/eslint-plugin-unicorn/pull/1

It handles the cached array length case in `unicorn/no-for-loop`, for example:

```js
for (let i = 0, j = arr.length; i < j; i += 1) {
	const element = arr[i];
}
```

The branch includes the rule change and a regression test in `test/no-for-loop.js`.

I tried to open an upstream PR, but GitHub reports that this repository only allows collaborators to open pull requests. The fork PR includes the validation notes and CI caveats from the fork run.

**Comment by a1local** [comment]
/attempt #295

Working on this in branch `bounty/no-for-loop-length-alias`.

**Comment by cedar323** [comment]
I have a focused fix branch for this here: https://github.com/cedar323/eslint-plugin-unicorn/tree/fix-no-for-loop-cached-length-295

It updates `no-for-loop` to detect the cached-length form from the issue while keeping autofix conservative when the cached length variable is used in the loop body or leaks outside. Local checks run: `node --check rules/no-for-loop.js`, `node --check test/no-for-loop.js`, and `git diff --check`.

I attempted to open the upstream PR from the fork, but the current GitHub token available in this environment is not permitted to create PRs against this repository.

**Comment by maikeru199** [comment]
I have a focused branch for this here:

https://github.com/maikeru199/eslint-plugin-unicorn/tree/codex/no-for-loop-cached-length

Compare view:

https://github.com/sindresorhus/eslint-plugin-unicorn/compare/main...maikeru199:eslint-plugin-unicorn:codex/no-for-loop-cached-length?expand=1

It updates `no-for-loop` to detect cached `.length` loop bounds such as `for (let i = 0, length = items.length; i < length; i++)`, reusing the existing `for...of` / `.entries()` fixer path when safe. It also avoids autofixing when the cached length variable is referenced outside the loop header, and includes regression coverage in `test/no-for-loop.js`.

Validation run locally:

- `git diff --check origin/main...HEAD`
- `node --check rules/no-for-loop.js`
- `node --check test/no-for-loop.js`

I tried to open an upstream PR from the fork, but GitHub shows: `An owner of this repository has limited the ability to open a pull request to users that are collaborators on this repository.`

Update 2026-06-02: I retried creating the PR through the GitHub API as well, and it still returns `403 Resource not accessible by integration`, so the collaborator-only PR restriction appears to still be in effect.

I also saw the earlier discussion about possibly handling this as a separate rule. If that is still the preferred direction, I can adapt the branch accordingly.

**Comment by sato820** [comment]
I have a tested patch for this issue in my fork:
https://github.com/sindresorhus/eslint-plugin-unicorn/compare/main...sato820:codex/no-for-loop-cached-length

GitHub shows that pull request creation is limited to collaborators on this repository, so I cannot open the PR directly from my fork. The branch extends `no-for-loop` to handle cached `array.length` loop bounds and includes tests for the conservative skip cases.

Validation run:
- `node ./node_modules/eslint/bin/eslint.js rules/no-for-loop.js test/no-for-loop.js`
- `node ./node_modules/ava/entrypoints/cli.mjs test/no-for-loop.js --timeout=120s --serial`
- `git diff --check`
